poly trajectory input (system2.0)

naveenoid opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3295)
<!-- Reviewable:end -->

amcastro-tri commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.cc, line 38 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KQOC2-3SGkOO4DqWX3i:-KQOC2-3SGkOO4DqWX3j:b-mx5w6s) ([raw file](https://github.com/robotlocomotion/drake/blob/7a14f3678e8dcd8dc40fba95b9a78ab37e032f1e/drake/systems/framework/primitives/time_varying_polynomial_source.cc#L38)):*
> ```cpp
>                                         SystemOutput<T>* output) const {
>   DRAKE_ASSERT(output->get_num_ports() == 1);
>   int time = context.get_time();
> ```

Notice that time should be of type `T` (in simulation would be `double` not `int`)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

nkoenig commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.cc, line 22 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KRjB4fj5dH6Y4n7jn85:-KRjB4fj5dH6Y4n7jn86:b-39zwep) ([raw file](https://github.com/robotlocomotion/drake/blob/7244bb797b6e47330240c196437e55706367e30a/drake/systems/framework/primitives/time_varying_polynomial_source.cc#L22)):*
> ```cpp
> void TimeVaryingPolynomialSource<T>::EvalOutput(const ContextBase<T>& context,
>                                         SystemOutput<T>* output) const {
>   DRAKE_ASSERT(output->get_num_ports() == 1);
> ```

Do you want this to be `DRAKE_ASSERT(output && output->get_num_ports() == 1);`?

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KRjBi_JjO_OZOXtDcm1:-KRjBi_JjO_OZOXtDcm2:b-xens09) ([raw file](https://github.com/robotlocomotion/drake/blob/7244bb797b6e47330240c196437e55706367e30a/drake/systems/framework/primitives/time_varying_polynomial_source.h#L35)):*
> ```objc
> 
>   /// Outputs a signal with a time-varying polynomial value as specified by the
>   /// user.
> ```

Is documentation of `context` and `output` required or needed here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.cc, line 22 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KRjB4fj5dH6Y4n7jn85:-KRjE5t1vbjaOlf9EYbW:boj1qgt) ([raw file](https://github.com/robotlocomotion/drake/blob/7244bb797b6e47330240c196437e55706367e30a/drake/systems/framework/primitives/time_varying_polynomial_source.cc#L22)):*
<details><summary><i>Previously, nkoenig (Nate Koenig) wrote…</i></summary>
> Do you want this to be `DRAKE_ASSERT(output && output->get_num_ports() == 1);`?
</details>
Good catch. Done.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KRjBi_JjO_OZOXtDcm1:-KRjCV8idHZu6kHKx3BC:b-qiu19) ([raw file](https://github.com/robotlocomotion/drake/blob/7244bb797b6e47330240c196437e55706367e30a/drake/systems/framework/primitives/time_varying_polynomial_source.h#L35)):*
<details><summary><i>Previously, nkoenig (Nate Koenig) wrote…</i></summary>
> Is documentation of `context` and `output` required or needed here?
</details>
Since this method overrides an inherited method, we generally don't require a new Doxygen comment.

In the past, we've been defecting partial documentation of parameters. In other words, either _all_ of the parameters are documented or _none_ of the parameters are documented.

Regarding whether documentation is needed, I'll leave that for @naveenoid to answer :smile: .

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/framework/primitives/time_varying_polynomial_source.cc, line 22 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KRjB4fj5dH6Y4n7jn85:-KRjFZcfvzelXuMmTWs1:b-h5vcdo) ([raw file](https://github.com/robotlocomotion/drake/blob/7244bb797b6e47330240c196437e55706367e30a/drake/systems/framework/primitives/time_varying_polynomial_source.cc#L22)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Good catch. Done.
</details>
Actually it should be `DRAKE_ASSERT_VOID(IsValidContext(context));` or however that's spelled.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.cc, line 22 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KRjB4fj5dH6Y4n7jn85:-KRjIBkqZWz-P28auyLG:boyexbg) ([raw file](https://github.com/robotlocomotion/drake/blob/7244bb797b6e47330240c196437e55706367e30a/drake/systems/framework/primitives/time_varying_polynomial_source.cc#L22)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Actually it should be `DRAKE_ASSERT_VOID(IsValidContext(context));` or however that's spelled.
</details>
Done. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Assigning a fresh set of eyes to pass this straightforward PR : 

+@ggould-tri for platform review
+@sammy-tri for feature review

naveenoid commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.cc, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KQOC2-3SGkOO4DqWX3i:-KSL5vmx0rgk1WRlBR2d:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7a14f3678e8dcd8dc40fba95b9a78ab37e032f1e/drake/systems/framework/primitives/time_varying_polynomial_source.cc#L38)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Notice that time should be of type `T` (in simulation would be `double` not `int`)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KRjBi_JjO_OZOXtDcm1:-KSXsm-vBxzACJfBku3r:b3dfi3y) ([raw file](https://github.com/robotlocomotion/drake/blob/7244bb797b6e47330240c196437e55706367e30a/drake/systems/framework/primitives/time_varying_polynomial_source.h#L35)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Since this method overrides an inherited method, we generally don't require a new Doxygen comment.
> 
> In the past, we've been defecting partial documentation of parameters. In other words, either _all_ of the parameters are documented or _none_ of the parameters are documented.
> 
> Regarding whether documentation is needed, I'll leave that for @naveenoid to answer :smile: .
</details>
I will not document at this point since as @liangfok points out, params are inherited. I can review if the reviewers think otherwise.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot retest this please

ggould-tri commented:
Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KRjBi_JjO_OZOXtDcm1:-KSbEQuBXkUt_fn0_nzc:b7shhei) ([raw file](https://github.com/robotlocomotion/drake/blob/7244bb797b6e47330240c196437e55706367e30a/drake/systems/framework/primitives/time_varying_polynomial_source.h#L35)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> I will not document at this point since as @liangfok points out, params are inherited. I can review if the reviewers think otherwise.
</details>
That is consistent with google style guide.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 29 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbFtSsX0XJxuQ4iEIX:-KSbFtSsX0XJxuQ4iEIY:bl659wx) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/primitives/time_varying_polynomial_source.h#L29)):*
> ```objc
>   /// Constructs a system with a vector output that is time-varying and equals
>   /// the value of the piecewise polynomial evaluated at a each time.
>   /// supplied
> ```

FYI -- Not clear to me what this word is doing here.  Was it meant to be in the previous sentence?

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 30 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbFoHd1RORs32hUuV0:-KSbFoHd1RORs32hUuV1:b-sa0vht) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/primitives/time_varying_polynomial_source.h#L30)):*
> ```objc
>   /// the value of the piecewise polynomial evaluated at a each time.
>   /// supplied
>   /// @param pp_traj PiecewisePolynomial used by the system. so that the output
> ```

FYI -- 'S' here should be capitalized as it begins a sentence.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 33 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbG54_umRc2CbrRtJZ:-KSbG54_umRc2CbrRtJ_:b-ay9peo) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/primitives/time_varying_polynomial_source.h#L33)):*
> ```objc
>   /// is `y = pp_traj(t)` at all times.
>   explicit TimeVaryingPolynomialSource(
>       const PiecewisePolynomial<double>& pp_traj);
> ```

The current implementation aliases this reference.  You must either document here that the supplied `pp_traj` must remain valid for the lifetime of the system or else update the implementation to copy the value rather than the reference.

---

*[drake/systems/framework/primitives/test/time_varying_polynomial_source_test.cc, line 25 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbEjtG5mMqyYqq4HcB:-KSbEjtG5mMqyYqq4HcC:b-q3x5uq) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/primitives/test/time_varying_polynomial_source_test.cc#L25)):*
> ```cpp
>       : kppTraj(Eigen::MatrixXd::Constant(2, 2, 1.5)) {}
>   void SetUp() override {
>     //    PiecewisePolynomial<double>& pp_traj
> ```

Commented out code should be explained or deleted.

---

*[drake/systems/framework/primitives/test/time_varying_polynomial_source_test.cc, line 37 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbGfYm7dxxYoJCQ2EL:-KSbGfYm7dxxYoJCQ2EM:bcccrmi) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/primitives/test/time_varying_polynomial_source_test.cc#L37)):*
> ```cpp
>   }
> 
>   const PiecewisePolynomial<double> kppTraj;
> ```

FYI -- Not clear why this constant belongs in this class; if it were outside of the class it could be reused instead of the `test_pp` copypasta below.  So consider moving it.

In addition, the letter following the `k` prefix is generally capitalized, so also consider the name `kPpTraj`.  I know that lowercase `pp` is common throughout this code, so it's your choice which way to go on this one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/framework/CMakeLists.txt, line 23 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbJ8jrksQM901fLYN8:-KSbJ8jwfKmFQpae60Rz:bh80y8t) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/CMakeLists.txt#L23)):*
> ```CMake
>   primitives/pass_through.cc
>   primitives/pid_controller.cc
>   primitives/time_varying_polynomial_source.cc
> ```

(apologies in advance for starting this)

Is there a known/preferred scope for what counts as a "primitive" for System2 vs what's just an implementation of a system?  I ask because this is the System2 implementation of drake/systems/controllers/open_loop_trajectory_controller.h.   I'm not arguing that controllers/ is the right place for this code, just trying to figure out how we could make such decisions going forward.

If we don't know, I guess I won't block on it...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/framework/CMakeLists.txt, line 23 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbJ8jrksQM901fLYN8:-KSbLssoLmyG5tHUgX6g:blkpfmb) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/CMakeLists.txt#L23)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> (apologies in advance for starting this)
> 
> Is there a known/preferred scope for what counts as a "primitive" for System2 vs what's just an implementation of a system?  I ask because this is the System2 implementation of drake/systems/controllers/open_loop_trajectory_controller.h.   I'm not arguing that controllers/ is the right place for this code, just trying to figure out how we could make such decisions going forward.
> 
> If we don't know, I guess I won't block on it...
</details>
Beetlejuicing @sherm1 or @david-german-tri for a System2 architecture view on this question.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/framework/CMakeLists.txt, line 23 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbJ8jrksQM901fLYN8:-KSbMATZCAH-fHvl-sRl:b-tar3h4) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/CMakeLists.txt#L23)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> Beetlejuicing @sherm1 or @david-german-tri for a System2 architecture view on this question.
</details>
BTW I specifically named this a primitive since I believe this is not really a controller but a signal source similar to the constant value/vector sources in this directory. I will wait for the architect's take on this.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 29 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbFtSsX0XJxuQ4iEIX:-KSbMhtJh4akFla6lDxI:b-wnfmk9) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/primitives/time_varying_polynomial_source.h#L29)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> FYI -- Not clear to me what this word is doing here.  Was it meant to be in the previous sentence?
</details>
Done. Removed.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 30 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbFoHd1RORs32hUuV0:-KSbTakX-EDCoF7gH9lh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/primitives/time_varying_polynomial_source.h#L30)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> FYI -- 'S' here should be capitalized as it begins a sentence.
</details>
Done.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 33 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbG54_umRc2CbrRtJZ:-KSbNN39DKJs8dEG6CPy:b52ezfy) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/primitives/time_varying_polynomial_source.h#L33)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> The current implementation aliases this reference.  You must either document here that the supplied `pp_traj` must remain valid for the lifetime of the system or else update the implementation to copy the value rather than the reference.
</details>
Done. Updated to copy the value instead.

---

*[drake/systems/framework/primitives/test/time_varying_polynomial_source_test.cc, line 25 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbEjtG5mMqyYqq4HcB:-KSbNXhEigAX5ywnkKmE:b-wnfmk9) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/primitives/test/time_varying_polynomial_source_test.cc#L25)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> Commented out code should be explained or deleted.
</details>
Done. Removed.

---

*[drake/systems/framework/primitives/test/time_varying_polynomial_source_test.cc, line 37 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbGfYm7dxxYoJCQ2EL:-KSbTfjK1Sm9swRnKB7D:bl498ux) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/primitives/test/time_varying_polynomial_source_test.cc#L37)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> FYI -- Not clear why this constant belongs in this class; if it were outside of the class it could be reused instead of the `test_pp` copypasta below.  So consider moving it.
> 
> In addition, the letter following the `k` prefix is generally capitalized, so also consider the name `kPpTraj`.  I know that lowercase `pp` is common throughout this code, so it's your choice which way to go on this one.
</details>
Its needed in order to initialise source_ which in turn needs the TimeVaryingPolynomialSource object. I instead remove the copy-pasta to simply access the variable from the test below since it is permitted by the TEST_F type method.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 3 of 6 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 40 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbXfK6tLBeAv1Mf099:-KSbXfK6tLBeAv1Mf09A:b-nc8iba) ([raw file](https://github.com/robotlocomotion/drake/blob/5cc9f63118383ca7a5fd667dc4c1eddf2c3ed1da/drake/systems/framework/primitives/time_varying_polynomial_source.h#L40)):*
> ```objc
> 
>  private:
>   const PiecewisePolynomial<double> kPpTraj;
> ```

The kConstant spelling is not typically used for constructor-initialized member data, even if it is constant for the lifetime of the class instance.

https://google.github.io/styleguide/cppguide.html#Constant_Names only calls for the use of kConstant when the value is fixed for the duration of the program.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
FYI: https://drake-cdash.csail.mit.edu/testDetails.php?test=2970963&build=126429

>  time_varying_polynomial_source_test: /home/ubuntu/workspace/experimental/a29e56a9/build/install/include/eigen3/Eigen/src/Core/DenseBase.h:257: void Eigen::DenseBase<Eigen::Block<Eigen::Matrix<double, -1, 1, 0, -1, 1>, -1, 1, false> >::resize(Index, Index) [Derived = Eigen::Block<Eigen::Matrix<double, -1, 1, 0, -1, 1>, -1, 1, false>]: Assertion `rows == this->rows() && cols == this->cols() && "DenseBase::resize() does not actually allow to resize."` failed.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-:-KSbxGwNwbcyIwS0ybFc:b-6a5vm2)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/CMakeLists.txt, line 23 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbJ8jrksQM901fLYN8:-KScAbNGFQxtSk8NGb3y:b-qp1dgv) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/CMakeLists.txt#L23)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> BTW I specifically named this a primitive since I believe this is not really a controller but a signal source similar to the constant value/vector sources in this directory. I will wait for the architect's take on this.
</details>
As a rule of thumb, it's a primitive if you can imagine it being used in at least two semantically unrelated projects.   For instance, automotive and manipulation.

I would say this passes that test.  It's a little niche, but it's not application-specific.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/CMakeLists.txt, line 23 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbJ8jrksQM901fLYN8:-KScEBZ6fOgUJlEfaOuW:b-8bxd3t) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/CMakeLists.txt#L23)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> As a rule of thumb, it's a primitive if you can imagine it being used in at least two semantically unrelated projects.   For instance, automotive and manipulation.
> 
> I would say this passes that test.  It's a little niche, but it's not application-specific.
</details>
I worry that could eventually include a lot of controller-ish functionality, but this is simple enough that I defer any objections for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/CMakeLists.txt, line 23 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbJ8jrksQM901fLYN8:-KSdj6hy0SslbBcHLy4J:b-13mm3h) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/CMakeLists.txt#L23)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> I worry that could eventually include a lot of controller-ish functionality, but this is simple enough that I defer any objections for now.
</details>
I like David's rule of thumb. I've been thinking of "primitive" here as analogous to Simulink's [block library](https://www.mathworks.com/help/simulink/block-libraries.html), which contains a broad assortment of useful blocks including basic controllers. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Fixed.. PiecewisePolynomial used in the test generated a problem.

---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/framework/CMakeLists.txt, line 23 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbJ8jrksQM901fLYN8:-KSgYIKm_e0G3saGtWmU:bt7b3cn) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/CMakeLists.txt#L23)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> I like David's rule of thumb. I've been thinking of "primitive" here as analogous to Simulink's [block library](https://www.mathworks.com/help/simulink/block-libraries.html), which contains a broad assortment of useful blocks including basic controllers. 
</details>
Done. (For now I guess nothing to address)

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 40 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbXfK6tLBeAv1Mf099:-KSg6WhD7jWhyaKTCtYC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5cc9f63118383ca7a5fd667dc4c1eddf2c3ed1da/drake/systems/framework/primitives/time_varying_polynomial_source.h#L40)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> The kConstant spelling is not typically used for constructor-initialized member data, even if it is constant for the lifetime of the class instance.
> 
> https://google.github.io/styleguide/cppguide.html#Constant_Names only calls for the use of kConstant when the value is fixed for the duration of the program.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-:-KSgY0EVn7yppQUtY6RI:bgqohvf)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 6 of 6 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.cc, line 8 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSgahf6VTYhP5n9tVEj:-KSgahf6VTYhP5n9tVEk:b-ifo69a) ([raw file](https://github.com/robotlocomotion/drake/blob/57b44a7698f408a8620a4f0f80ce3721adbe38d6/drake/systems/framework/primitives/time_varying_polynomial_source.cc#L8)):*
> ```cpp
> #include "drake/common/drake_assert.h"
> #include "drake/systems/framework/context.h"
> #include<iostream>
> ```

This is unused, and is missing a space, and probably shouldn't be here at all.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 6 of 6 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/framework/primitives/test/time_varying_polynomial_source_test.cc, line 22 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSgk_-vO7e2vz8xbsV_:-KSgk_-w3VUXowjcnN9-:b-tg8elf) ([raw file](https://github.com/robotlocomotion/drake/blob/57b44a7698f408a8620a4f0f80ce3721adbe38d6/drake/systems/framework/primitives/test/time_varying_polynomial_source_test.cc#L22)):*
> ```cpp
> class TimeVaryingPolynomialSourceTest : public ::testing::Test {
>  protected:
>   TimeVaryingPolynomialSourceTest() :
> ```

https://google.github.io/styleguide/cppguide.html#Constructor_Initializer_Lists shows the : character on the line with the variable being initialized.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: 4 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.cc, line 8 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSgahf6VTYhP5n9tVEj:-KSggtp4x3g-AjkEq9-P:b-7xomxp) ([raw file](https://github.com/robotlocomotion/drake/blob/57b44a7698f408a8620a4f0f80ce3721adbe38d6/drake/systems/framework/primitives/time_varying_polynomial_source.cc#L8)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> This is unused, and is missing a space, and probably shouldn't be here at all.
</details>
OMG..was running some tests an it sneaked in. Removed.

---

*[drake/systems/framework/primitives/test/time_varying_polynomial_source_test.cc, line 22 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSgk_-vO7e2vz8xbsV_:-KSgm2-6zUbJ1xhtSR8J:bs7wumx) ([raw file](https://github.com/robotlocomotion/drake/blob/57b44a7698f408a8620a4f0f80ce3721adbe38d6/drake/systems/framework/primitives/test/time_varying_polynomial_source_test.cc#L22)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> https://google.github.io/styleguide/cppguide.html#Constructor_Initializer_Lists shows the : character on the line with the variable being initialized.
</details>
Done. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-:-KSgn0660K_8GizrQlhD:bnfp4nl)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-:-KSgo5-wtgEfj449TI2b:bnfp4nl)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/framework/CMakeLists.txt, line 23 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KSbJ8jrksQM901fLYN8:-KSk8mIXIurH1nDYcVka:b-vckabk) ([raw file](https://github.com/robotlocomotion/drake/blob/3bab837c027101da120b49caef63bef1e1f0c229/drake/systems/framework/CMakeLists.txt#L23)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> Done. (For now I guess nothing to address)
</details>
Marking this discussion as resolved..

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.cc, line 22 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KRjB4fj5dH6Y4n7jn85:-KSk8uieIHTrwS2GgcPi:b-uaff4a) ([raw file](https://github.com/robotlocomotion/drake/blob/7244bb797b6e47330240c196437e55706367e30a/drake/systems/framework/primitives/time_varying_polynomial_source.cc#L22)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Done. Thanks.
</details>
Can i mark this discussion as resolved?

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KRjBi_JjO_OZOXtDcm1:-KSk8zY6i-u5mqKxp2OR:b-eweg8j) ([raw file](https://github.com/robotlocomotion/drake/blob/7244bb797b6e47330240c196437e55706367e30a/drake/systems/framework/primitives/time_varying_polynomial_source.h#L35)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> That is consistent with google style guide.
</details>
Is it ok to mark this as resolved?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.cc, line 22 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KRjB4fj5dH6Y4n7jn85:-KSkdbGM9DEFWlVawIrz:bg7igkv) ([raw file](https://github.com/robotlocomotion/drake/blob/7244bb797b6e47330240c196437e55706367e30a/drake/systems/framework/primitives/time_varying_polynomial_source.cc#L22)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> Can i mark this discussion as resolved?
</details>
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KRjBi_JjO_OZOXtDcm1:-KSkdtVzrT4znfjOK3wj:bg7igkv) ([raw file](https://github.com/robotlocomotion/drake/blob/7244bb797b6e47330240c196437e55706367e30a/drake/systems/framework/primitives/time_varying_polynomial_source.h#L35)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> Is it ok to mark this as resolved?
</details>
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.cc, line 22 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KRjB4fj5dH6Y4n7jn85:-KSkoVedqOW0iQQhmki2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7244bb797b6e47330240c196437e55706367e30a/drake/systems/framework/primitives/time_varying_polynomial_source.cc#L22)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> :lgtm: 
</details>
Done.

---

*[drake/systems/framework/primitives/time_varying_polynomial_source.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-KRjBi_JjO_OZOXtDcm1:-KSkoY8FenPyOW2C4SFz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7244bb797b6e47330240c196437e55706367e30a/drake/systems/framework/primitives/time_varying_polynomial_source.h#L35)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> :lgtm: 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> apparently I am an assignee! 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3295#-:-KSkqQCYaXQC125cKT8E:b-vhvv6d)*
<!-- Sent from Reviewable.io -->
